lib/Basics/GlobalSerialization.cpp

```diff
@@ -0,0 +1,145 @@
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Max Neunhoeffer
+////////////////////////////////////////////////////////////////////////////////
+
+#include "Basics/GlobalSerialization.h"
+
+#include <thread>
+
+#include "Basics/FileUtils.h"
+#include "Basics/StringUtils.h"
+#include "Basics/files.h"
+#include "Logger/LogMacros.h"
+
+namespace arangodb {
+
+namespace {
+std::pair<std::string, std::string> findSLPProgramPaths() {
+  std::string path = "/tmp";
+  TRI_GETENV("ARANGOTEST_ROOT_DIR", path);
+  std::string progPath =
+      basics::FileUtils::buildFilename(path.c_str(), "globalSLP");
+  std::string pcPath =
+      basics::FileUtils::buildFilename(path.c_str(), "globalSLP_PC");
+  return std::pair(progPath, pcPath);
+}
+
+std::vector<std::string> readSLPProgram(std::string const& progPath) {
+  // Read program:
+  std::vector<std::string> progLines;
+  try {
+    std::string prog = arangodb::basics::FileUtils::slurp(progPath);
+    if (prog.empty()) {
+      return progLines;
+    }
+    progLines = arangodb::basics::StringUtils::split(prog, '\n');
+    progLines.pop_back();  // Ignore last line, since every line is
+                           // ended by \n.
+  } catch (std::exception const&) {
+    // No program, return empty
+  }
+  return progLines;
+}
+
+}  // namespace
+
+void waitForGlobalEvent(std::string_view id, std::string_view selector) {
+  auto [progPath, pcPath] = findSLPProgramPaths();
+  std::vector<std::string> progLines = readSLPProgram(progPath);
+  if (progLines.empty()) {
+    return;
+  }
+  LOG_TOPIC("ace32", INFO, Logger::MAINTENANCE)
+      << "Waiting for global event " << id << " with selector " << selector
+      << "...";
+  while (true) {
+    // Read pc
+    std::vector<std::string> executedLines = readSLPProgram(pcPath);
+    if (executedLines.size() >= progLines.size()) {
+      return;  // program already finished
+    }
+    std::string current = progLines[executedLines.size()];
+    std::vector<std::string> parts =
+        arangodb::basics::StringUtils::split(current, ' ');
+    if (parts.size() >= 2 && id == parts[0] && selector == parts[1]) {
+      // Hurray! We can make progress:
+      if (parts.size() >= 3) {
+        LOG_TOPIC("ace33", INFO, Logger::MAINTENANCE)
+            << "Global event " << id << " with selector " << selector
+            << " and comment " << parts[2] << " has happened...";
+      } else {
+        LOG_TOPIC("ace34", INFO, Logger::MAINTENANCE)
+            << "Global event " << id << " with selector " << selector
+            << " has happened...";
+      }
+      current.push_back('\n');
+      try {
+        arangodb::basics::FileUtils::appendToFile(pcPath, current);
```

Comment:
If multiple threads of the same arangod instance, or even different arangod instances write to the same `globalSLP_PC` file, this could be unsafe. In theory it is possible that one writer instance writes a line only partially, and then another writer kicks in. That is at least possible if multiple writers can act upon the same trigger concurrently. 
Not sure if this is actually possible in the test.

Reply:
Currently it is not possible. This is, because the file is only ever modified, if we are at a specific code place which matches the current line. So it is only one code place which can modify the file in a given state. If we can now ensure that this code place is not executed by two threads concurrently, we are good. As far as I see this is currently the case.
Secondly, the code only looks at the **number of lines** in the file, that is the number of '\n' bytes and this should be the same, even if lines are mixed, but this is of course nothing we can count on.
Therefore I completely agree that this is dangerous to be misused. Therefore, I suggest to put a global mutex into this code, ensuring that only one thread at a time calls `appendToFile` on this file.
We then have "only" the problem left that in the same cluster two such tests could be executed at the same time, which is inherent in the design of this system, and I currently do not see a way around this except if we would hand in the global name of the file into the cluster for each test, which I think is overkill and overcomplicated.
Therefore I suggest to live with this problem. I will add a loud warning in the header file.

Reply:
Note that the path to the SLP program does contain the test-suite (e.g. shell-client), which somewhat limits the problem. And since we do not execute multiple such tests in the suite in parallel, we should be good.